File: cloudflare/src/pipeline/scorer.ts
import { NormalizedItem, GoalRow } from '../types';

export function scoreItems(items: NormalizedItem[], goals: GoalRow[]): NormalizedItem[] {
  const goalCategories = goals.map((g) =>
    new Set((g.keywords ?? []).map((k) => k.category))
  );

  for (const item of items) {
    item.quality_score = qualityScore(item);
    item.signal_score = signalScore(item, goalCategories);
  }

  return items;
}

function qualityScore(item: NormalizedItem): number {
  let score = 0;
  const text = item.text_content ?? '';
  const eng = item.engagement_snapshot ? JSON.parse(item.engagement_snapshot) : null;

  if (text.length > 10) score += 0.2;
  if (text.length > 100) score += 0.15;
  if (item.title) score += 0.15;
  if (eng && Object.values(eng).some((v) => typeof v === 'number' && v > 0)) score += 0.2;
  if (item.publish_time) {
    const age = Date.now() - new Date(item.publish_time).getTime();
    if (age < 86_400_000) score += 0.15; // < 24h
  }
  if (!item.is_duplicate) score += 0.15;

  return Math.round(Math.min(score, 1) * 1000) / 1000;
}

function signalScore(item: NormalizedItem, goalCategories: Set<string>[]): number {
  if (!goalCategories.length) return 0.1;

  const tags = new Set<string>(JSON.parse(item.tags));
  if (!tags.size) return 0.1;

  const matches = goalCategories.filter((cats) =>
    [...tags].some((t) => cats.has(t))
  ).length;

  if (matches === 0) return 0.1;
  const max = goalCategories.length;
  return Math.round(Math.min(0.4 + 0.6 * (matches - 1) / Math.max(max - 1, 1), 1) * 1000) / 1000;
}

File: src/pipeline/scorer.py
from __future__ import annotations

from datetime import datetime, timezone, timedelta
from typing import Any

from src.models.goal import Goal


def score_items(
    items: list[dict[str, Any]],
    goals: list[Goal],
) -> list[dict[str, Any]]:
    """Score items for quality and signal relevance.

    Quality score (0-1): weighted sum of content richness signals.
    Signal score (0-1): relevance to active research goals.
    """
    # Build goal keyword sets for signal scoring
    goal_categories: list[set[str]] = []
    for goal in goals:
        cats = set()
        for kw in (goal.keywords or []):
            cats.add(kw.category)
        goal_categories.append(cats)

    now = datetime.now(timezone.utc)

    for item in items:
        item["quality_score"] = _quality_score(item, now)
        item["signal_score"] = _signal_score(item, goal_categories)

    return items


def _quality_score(item: dict[str, Any], now: datetime) -> float:
    """Calculate quality score based on content richness."""
    score = 0.0
    text = item.get("text_content") or ""
    title = item.get("title") or ""
    engagement = item.get("engagement_snapshot")
    publish_time = item.get("publish_time")
    is_dup = item.get("is_duplicate", False)

    # Has meaningful text (0.2)
    if len(text) > 10:
        score += 0.2

    # Text length bonus (0.15)
    if len(text) > 100:
        score += 0.15

    # Has title (0.15)
    if title:
        score += 0.15

    # Has engagement data (0.2)
    if engagement and any(v for v in engagement.values() if isinstance(v, (int, float)) and v > 0):
        score += 0.2

    # Recency bonus (0.15) - within last 24 hours
    if publish_time and isinstance(publish_time, datetime):
        if (now - publish_time) < timedelta(hours=24):
            score += 0.15

    # Not duplicate (0.15)
    if not is_dup:
        score += 0.15

    return round(min(score, 1.0), 3)


def _signal_score(item: dict[str, Any], goal_categories: list[set[str]]) -> float:
    """Calculate signal score based on goal relevance."""
    if not goal_categories:
        return 0.1  # Baseline when no goals defined

    item_tags = set(item.get("tags") or [])
    if not item_tags:
        return 0.1

    # Score based on how many goals this item's tags match
    matches = 0
    for goal_cats in goal_categories:
        if item_tags & goal_cats:
            matches += 1

    if matches == 0:
        return 0.1

    # Scale: matching 1 goal = 0.4, all goals = 1.0
    return round(min(0.4 + (0.6 * (matches - 1) / max(len(goal_categories) - 1, 1)), 1.0), 3)
